streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -720,50 +695,72 @@ private void closeGlobalStreamThread() {
      * Note that this method must not be called in the {@code onChange} callback of {@link StateListener}.
      */
     public synchronized boolean close(final long timeout, final TimeUnit timeUnit) {
-        log.debug("{} Stopping Kafka Stream process.", logPrefix);
+        log.debug("{} Stopping Streams client", logPrefix);
 
-        // only clean up once
-        if (!checkFirstTimeClosing()) {
-            return true;
-        }
+        if (!setState(State.PENDING_SHUTDOWN)) {
+            // if transition failed, it means it was either in PENDING_SHUTDOWN
+            // or NOT_RUNNING already; return immediately in this case
+            log.info("{} Already stopped, do not proceed with the shutdown process", logPrefix);
 
-        stateDirCleaner.shutdownNow();
-        // save the current thread so that if it is a stream thread
-        // we don't attempt to join it and cause a deadlock
-        final Thread shutdown = new Thread(new Runnable() {
-            @Override
-            public void run() {
-                // signal the threads to stop and wait
-                for (final StreamThread thread : threads) {
-                    // avoid deadlocks by stopping any further state reports
-                    // from the thread since we're shutting down
-                    thread.setStateListener(null);
-                    thread.close();
-                }
-                closeGlobalStreamThread();
-                for (final StreamThread thread : threads) {
-                    try {
-                        if (!thread.stillRunning()) {
+            return true;
```

Comment:
I'm not sure returning `true` is valid. We don't actually know if all the threads have shutdown. Though, i'm not entirely sure what to do about it. Perhaps we need to extract the shutdown Thread as a field and then we can check if it is still running. If it isn't running then we can return true, otherwise we should try and join on the thread with the provided timeout